src/lib.rs

```diff
@@ -124,96 +24,1138 @@ impl<T> RpcMessage for T where
+    /// In the remote case, this involves lazily creating a connection to the
+    /// remote side and then creating a new stream on the underlying
+    /// [`quinn`] or iroh connection.
+    ///
+    /// In both cases, the returned sender is fully self contained.
+    #[allow(clippy::type_complexity)]
+    pub fn request(
+        &self,
+    ) -> impl Future<
+        Output = Result<Request<LocalSender<M, S>, rpc::RemoteSender<R, S>>, RequestError>,
+    > + 'static
+    where
+        S: Service,
+        M: Send + Sync + 'static,
+        R: 'static,
+    {
+        #[cfg(feature = "rpc")]
+        {
+            let cloned = match &self.0 {
+                ClientInner::Local(tx) => Request::Local(tx.clone()),
+                ClientInner::Remote(connection) => Request::Remote(connection.clone_boxed()),
+            };
+            async move {
+                match cloned {
+                    Request::Local(tx) => Ok(Request::Local(tx.into())),
+                    #[cfg(feature = "rpc")]
+                    Request::Remote(conn) => {
+                        let (send, recv) = conn.open_bi().await?;
+                        Ok(Request::Remote(rpc::RemoteSender::new(send, recv)))
+                    }
+                }
+            }
+        }
+        #[cfg(not(feature = "rpc"))]
+        {
+            let ClientInner::Local(tx) = &self.0 else {
+                unreachable!()
+            };
+            let tx = tx.clone().into();
+            async move { Ok(Request::Local(tx)) }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(crate) enum ClientInner<M> {
+    Local(tokio::sync::mpsc::Sender<M>),
+    #[cfg(feature = "rpc")]
+    #[cfg_attr(quicrpc_docsrs, doc(cfg(feature = "rpc")))]
+    Remote(Box<dyn rpc::RemoteConnection>),
+    #[cfg(not(feature = "rpc"))]
+    #[cfg_attr(quicrpc_docsrs, doc(cfg(feature = "rpc")))]
+    Remote(PhantomData<M>),
+}
+
+impl<M> Clone for ClientInner<M> {
+    fn clone(&self) -> Self {
+        match self {
+            Self::Local(tx) => Self::Local(tx.clone()),
+            #[cfg(feature = "rpc")]
+            Self::Remote(conn) => Self::Remote(conn.clone_boxed()),
+            #[cfg(not(feature = "rpc"))]
+            Self::Remote(_) => unreachable!(),
+        }
+    }
+}
+
+/// Error when opening a request. When cross-process rpc is disabled, this is
+/// an empty enum since local requests can not fail.
+#[derive(Debug, thiserror::Error)]
+pub enum RequestError {
+    #[cfg(feature = "rpc")]
+    #[cfg_attr(quicrpc_docsrs, doc(cfg(feature = "rpc")))]
+    #[error("error establishing connection: {0}")]
+    Connect(#[from] quinn::ConnectError),
+    #[cfg(feature = "rpc")]
+    #[cfg_attr(quicrpc_docsrs, doc(cfg(feature = "rpc")))]
+    #[error("error opening stream: {0}")]
+    Connection(#[from] quinn::ConnectionError),
+    #[cfg(feature = "rpc")]
+    #[cfg_attr(quicrpc_docsrs, doc(cfg(feature = "rpc")))]
+    #[error("error opening stream: {0}")]
+    Other(#[from] anyhow::Error),
```

Comment:
maybe using an `io::Error` would match better with the rest of the system?

Reply:
The problem is that iroh currently is using anyhow::Error for connecting.

Connect returns an anyhow::Result, here: https://docs.rs/iroh/latest/iroh/endpoint/struct.Endpoint.html#method.connect

And Endpoint::open_bi also returns an anyhow::Result:
https://docs.rs/iroh/latest/iroh/endpoint/struct.OpenBi.html

I can of course use io::Error::other(theanyhowrerror). Not sure what our plan is there for iroh 1.0. What we don't want here is to put in a concrete iroh error, because then we have the semicircular dependency again.